File: problem-3/problem-3.go
package main

import (
	"fmt"
	"math"
	"slices"
)

func main() {
	const given_composite = 600851475143
	composite_sqrt := int(math.Sqrt(float64(given_composite))) + 1

	// Create prime sieve
	prime_sieve := make([]bool, composite_sqrt)

	// Populate the prime sieve to assume all numbers are prime
	for i := range composite_sqrt {
		prime_sieve[i] = true
	}

	// Set all composite numbered indexes in the sieve to false
	i := 2
	for i < composite_sqrt {
		// Set all
		for index := i * i; index < composite_sqrt; index += i {
			prime_sieve[index] = false
		}

		// Update i to the next prime exposed by the sieve
		i++
		for i < composite_sqrt && !prime_sieve[i] {
			i++
		}
	}

	// Iterate sieve in reverse and mod the given composite until it equals 0
	for i, val := range slices.Backward(prime_sieve) {
		if val && given_composite%i == 0 {
			fmt.Println(i)
			break
		}
	}
}

File: problem-3/problem-3.py
import math

GIVEN_COMPOSITE = 600851475143
composite_sqrt = int(math.sqrt(GIVEN_COMPOSITE)) + 1

# Create prime sieve and populate to assume all numbers are prime, initially
prime_sieve = [True] * composite_sqrt

i = 2
while i < composite_sqrt:
    for index in range(i*i, composite_sqrt, i):
        prime_sieve[index] = False

    # Update i to the next prime exposed by the sieve
    i += 1
    while i < composite_sqrt and not prime_sieve[i]:
        i += 1

i = composite_sqrt -1
for value in reversed(prime_sieve):
    if value and GIVEN_COMPOSITE % i == 0:
        print(i)
        break
    i -= 1
